llap-tez/src/java/org/apache/hadoop/hive/llap/tezplugins/LlapTaskSchedulerService.java

```diff
@@ -1817,8 +1830,10 @@ protected void schedulePendingTasks() throws InterruptedException {
         Iterator<TaskInfo> taskIter = taskListAtPriority.iterator();
         boolean scheduledAllAtPriority = true;
         while (taskIter.hasNext()) {
-          // TODO Optimization: Add a check to see if there's any capacity available. No point in
-          // walking through all active nodes, if they don't have potential capacity.
+          // Early exit where there are no slots available
+          if (isCapacityFull.get()) {
```

Comment:
This can be outer if condition? only if cluster capacity is available run the task iterator.